lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -192,6 +224,63 @@ constFoldFIRRTLBinaryOp(Operation *op, ArrayRef<Attribute> operands,
   return getIntAttr(resultType, resultValue);
 }
 
+/// Applies the canonicalization function `canonicalize` to the given operation.
+///
+/// Determines which (if any) of the operation's operands are constants, and
+/// provides them as arguments to the callback function. Any `invalidvalue` in
+/// the input is mapped to a constant zero. The value returned from the callback
+/// is used as the replacement for `op`, and an additional pad operation is
+/// inserted if necessary. Does nothing if the result of `op` is of unknown
+/// width, in which case the necessity of a pad cannot be determined.
+static LogicalResult canonicalizePrimOp(
+    Operation *op, PatternRewriter &rewriter,
+    const function_ref<OpFoldResult(ArrayRef<Attribute>)> &canonicalize) {
+  // Can only operate on FIRRTL primary operations.
```

Comment:
Nit: I think this should be "primitive" and not "primary"?

Reply:
Yeah totally, fixed.